src/solver.rs

```diff
@@ -167,17 +167,7 @@ fn recursive_join(
         // its value must match the candidate. This *should* already be true
         // because we selected the bucket using the shared vars—but keep this
         // in case upstream bucketing logic ever changes.
-        let mut compatible = true;
-        for (k, v) in cand.iter() {
-            if *k == WORD_SENTINEL {
-                continue; // ignore the “whole word” sentinel binding
-            }
-            if let Some(prev) = env.get(k) && prev != v {
-                compatible = false;
-                break;
-            }
-        }
-        if !compatible {
+        if cand.iter().filter(|(k, _)| **k != WORD_SENTINEL).any(|(k, v)| env.get(k).is_some_and(|prev| prev != v)) {
```

Comment:
note that `any` short circuits ([doc](<https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any>))